spanner/metrics.go

```diff
@@ -0,0 +1,474 @@
+			attribute.String(metricLabelKeyDatabase, database),
+			attribute.String(metricLabelKeyClientUID, clientUID),
+			attribute.String(metricLabelKeyClientName, clientName),
+			// Skipping instance config until we have a way to get it
+			attribute.String(monitoredResLabelKeyInstanceConfig, "unknown"),
+			attribute.String(monitoredResLabelKeyLocation, detectClientLocation(ctx)),
+		},
+		shutdown: func() {},
+	}
+
+	tracerFactory.isDirectPathEnabled = false
+
+	var meterProvider *sdkmetric.MeterProvider
+	if metricsProvider == nil {
+		// Create default meter provider
+		mpOptions, err := builtInMeterProviderOptions(project)
+		if err != nil {
+			return tracerFactory, err
+		}
+		meterProvider = sdkmetric.NewMeterProvider(mpOptions...)
+
+		tracerFactory.enabled = true
+		tracerFactory.shutdown = func() { meterProvider.Shutdown(ctx) }
+	} else {
+		switch metricsProvider.(type) {
+		case noop.MeterProvider:
+			tracerFactory.enabled = false
+			return tracerFactory, nil
+		default:
+			tracerFactory.enabled = false
+			return tracerFactory, errors.New("unknown MetricsProvider type")
+		}
+	}
+
+	// Create meter and instruments
+	meter := meterProvider.Meter(builtInMetricsMeterName, metric.WithInstrumentationVersion(internal.Version))
+	err = tracerFactory.createInstruments(meter)
+	return tracerFactory, err
+}
+
+func builtInMeterProviderOptions(project string) ([]sdkmetric.Option, error) {
+	defaultExporter, err := newMonitoringExporter(context.Background(), project, exporterOpts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return []sdkmetric.Option{sdkmetric.WithReader(
+		sdkmetric.NewPeriodicReader(
+			defaultExporter,
+			sdkmetric.WithInterval(defaultSamplePeriod),
+		),
+	)}, nil
+}
+
+func (tf *builtinMetricsTracerFactory) createInstruments(meter metric.Meter) error {
+	var err error
+
+	// Create operation_latencies
+	tf.operationLatencies, err = meter.Float64Histogram(
+		nativeMetricsPrefix+metricNameOperationLatencies,
+		metric.WithDescription("Total time until final operation success or failure, including retries and backoff."),
+		metric.WithUnit(metricUnitMS),
+		metric.WithExplicitBucketBoundaries(bucketBounds...),
+	)
+	if err != nil {
+		return err
+	}
+
+	// Create attempt_latencies
+	tf.attemptLatencies, err = meter.Float64Histogram(
+		nativeMetricsPrefix+metricNameAttemptLatencies,
+		metric.WithDescription("Client observed latency per RPC attempt."),
+		metric.WithUnit(metricUnitMS),
+		metric.WithExplicitBucketBoundaries(bucketBounds...),
+	)
+	if err != nil {
+		return err
+	}
+
+	// Create operation_count
+	tf.operationCount, err = meter.Int64Counter(
+		nativeMetricsPrefix+metricNameOperationCount,
+		metric.WithDescription("The count of database operations."),
+		metric.WithUnit(metricUnitCount),
+	)
+
+	// Create attempt_count
+	tf.attemptCount, err = meter.Int64Counter(
+		nativeMetricsPrefix+metricNameAttemptCount,
+		metric.WithDescription("The number of additional RPCs sent after the initial attempt."),
```

Comment:
It's a bit confusing about the concept of `attempt`. Here, we exclude the initial attempt so the `initial attempt` is not actually a `attempt` here. I feel `attempt` more sounds like `how many retries we have done`.

Reply:
Updated the comment, attempt will include initial attempt too.